node/actors/executor/src/lib.rs

```diff
@@ -0,0 +1,207 @@
+//! Library files for the executor. We have it separate from the binary so that we can use these files in the tools crate.
+
+use crate::io::Dispatcher;
+use anyhow::Context as _;
+use concurrency::{ctx, ctx::channel, net, scope};
+use consensus::Consensus;
+use roles::{node, validator, validator::FinalBlock};
+use std::{mem, sync::Arc};
+use storage::{FallbackReplicaStateStore, ReplicaStateStore, WriteBlockStore};
+use sync_blocks::SyncBlocks;
+use utils::pipe;
+
+mod config;
+mod io;
+mod metrics;
+#[cfg(test)]
+mod tests;
+
+pub use self::config::{ConsensusConfig, ExecutorConfig, GossipConfig};
+
+/// Validator-related part of [`Executor`].
+#[derive(Debug)]
+struct ValidatorExecutor {
+    /// Consensus network configuration.
+    config: ConsensusConfig,
+    /// Validator key.
+    key: validator::SecretKey,
+    /// Store for replica state.
+    replica_state_store: Arc<dyn ReplicaStateStore>,
+    /// Sender of blocks finalized by the consensus algorithm.
+    blocks_sender: channel::UnboundedSender<FinalBlock>,
+}
+
+impl ValidatorExecutor {
+    /// Returns consensus network configuration.
+    fn consensus_config(&self) -> network::consensus::Config {
+        network::consensus::Config {
+            // Consistency of the validator key has been verified in constructor.
+            key: self.key.clone(),
+            public_addr: self.config.public_addr,
+        }
+    }
+}
+
+/// Executor allowing to spin up all actors necessary for a consensus node.
+#[derive(Debug)]
+pub struct Executor<S> {
+    /// General-purpose executor configuration.
+    executor_config: ExecutorConfig,
+    /// Secret key of the node.
+    node_key: node::SecretKey,
+    /// Block and replica state storage used by the node.
+    storage: Arc<S>,
+    /// Validator-specific node data.
+    validator: Option<ValidatorExecutor>,
+}
+
+impl<S: WriteBlockStore + 'static> Executor<S> {
+    /// Creates a new executor with the specified parameters.
+    pub fn new(
+        node_config: ExecutorConfig,
+        node_key: node::SecretKey,
+        storage: Arc<S>,
+    ) -> anyhow::Result<Self> {
+        anyhow::ensure!(
+            node_config.gossip.key == node_key.public(),
+            "config.gossip.key = {:?} doesn't match the secret key {:?}",
+            node_config.gossip.key,
+            node_key
+        );
+
+        Ok(Self {
+            executor_config: node_config,
+            node_key,
+            storage,
+            validator: None,
+        })
+    }
+
+    /// Sets validator-related data for the executor.
+    pub fn set_validator(
+        &mut self,
+        config: ConsensusConfig,
+        key: validator::SecretKey,
+        replica_state_store: Arc<dyn ReplicaStateStore>,
+        blocks_sender: channel::UnboundedSender<FinalBlock>,
```

Comment:
These options could be made more fluent (e.g., if `S` implements `ReplicaStateStore`, it doesn't need to be provided again, and "standard" `blocks_sender`s could be considered (e.g., one that saves a block to the store, or one that does nothing). I've decided against this to make code simpler.